prime-router/src/main/kotlin/fhirengine/utils/HL7Reader.kt

```diff
@@ -18,33 +23,122 @@ import java.util.Date
+     * organizations.  For example, the NIST profile for v251 includes fields that are only available in the v27
+     * standard spec.  To get around this fact, we take advantage that the specs are mostly backwards compatible;
+     * a NIST v251 can be parsed using the v271 structure successfully and will now also include the data from the
+     * fields only available in the standard v27.  The only caveat to this approach is that the HAPI library itself
+     * is not 100% backwards compatible.  A common error is that a v251 message will specify a component is a CE, but
+     * the v27 spec says it must be a CWE; though these two data types are compatible from a field standpoint, the HAPI
+     * library will throw a type error along the lines of "a CWE field cannot be set to a CE type".  To get around this
+     * issue, if the message cannot be parsed to v27 we fall back to parsing it as a v251 message.
+     *
+     *
+     * If no message models are returned by [getMessageModelClasses], the string is parsed using the default behavior
+     * of [Hl7InputStreamMessageIterator].
+     *
+     *
      * @return one or more HL7 messages
      * @throws IllegalArgumentException if the raw data cannot be parsed or no messages were read
      */
     fun getMessages(rawMessage: String): List<Message> {
+        val messageModelsToTry = getMessageModelClasses(rawMessage)
         val messages: MutableList<Message> = mutableListOf()
         if (rawMessage.isBlank()) {
             actionLogger.error(InvalidReportMessage("Provided raw data is empty."))
-        } else {
+        } else if (messageModelsToTry.isEmpty()) {
             try {
                 val iterator = Hl7InputStreamMessageIterator(rawMessage.byteInputStream())
                 while (iterator.hasNext()) {
                     messages.add(iterator.next())
                 }
-                // NOTE for batch hl7; should we be doing anything with the BHS and other headers
             } catch (e: Hl7InputStreamMessageStringIterator.ParseFailureError) {
                 logHL7ParseFailure(e)
             }
+        } else {
+            val validationContext = ValidationContextFactory.noValidation()
+            var parseError: Hl7InputStreamMessageStringIterator.ParseFailureError? = null
+            run modelLoop@{
+                messageModelsToTry.forEach { model ->
+                    val context = DefaultHapiContext(CanonicalModelClassFactory(model))
+                    context.validationContext = validationContext
+                    try {
+                        val iterator = Hl7InputStreamMessageIterator(rawMessage.byteInputStream(), context)
+                        while (iterator.hasNext()) {
+                            messages.add(iterator.next())
+                        }
+                    } catch (e: Hl7InputStreamMessageStringIterator.ParseFailureError) {
+                        parseError = e
```

Comment:
Just thinking about an error case here, what if we get here and `messages` isn't empty? Would happen if we could parse some messages with the given model, but not all of them (not sure if that's even possible, but I think each message is parsed by `iterator.next()`). Seems like maybe we'd want to clear out `messages` so that we can continue onto the next model in the list or appropriately log the failure if there are no more models to try.

Reply:
Clearing out the messages makes a ton of sense to me.